madgraph/interface/reweight_interface.py

```diff
@@ -1024,58 +1313,258 @@ def change_kinematics(self, event):
+                    state = True
+                comb_i = fks_common.FKSLeg({'id': buff_event[min_i].pid,'number': min_i+1,'state': state})
+                if min_j<= 1:
+                    state = False
+                else:
+                    state = True
+                comb_j = fks_common.FKSLeg({'id': buff_event[min_j].pid,'number': min_j+1,'state': state})
+                ij_comb =fks_common.combine_ij(comb_i,comb_j, self.model, dict={},pert='QCD')
+                if ij_comb == []:
+                    ij_comb =fks_common.combine_ij(comb_j,comb_i, self.model, dict={},pert='QCD')
+        
+                # For n+1-body reweighting
+                if min_inv > sud_cut:
+                    event_to_sud = buff_event
+                    n_part = nexternal+1
+                    mapped_tag, mapped_order = event_to_sud.get_tag_and_order()
+                    type = 1  #### H1 type
+
+                # For n-body reweighting
+                else:
+                    # If no reasonable recbination found, still use the n+1-body kinematics for sudakov
+                    if ij_comb == []:
+                        event_to_sud = buff_event
+                        n_part = nexternal+1
+                        mapped_tag, mapped_order = event_to_sud.get_tag_and_order()
+                        #### H1 type
+                        type = 1
+                    else:
+                        event_for_sud = self.merge_particles_kinematics(buff_event, min_i,min_j,ij_comb)
+                        event_to_sud = event_for_sud
+                        n_part = nexternal 
+                        mapped_tag, mapped_order = event_to_sud.get_tag_and_order()
+                        # map to n+1 body if recoil does not exist at Born level among processes
+                        if mapped_tag not in sud_mod.pdg2ewsud_dict.keys():
+                            event_to_sud = buff_event
+                            n_part = nexternal+1
+                            mapped_tag, mapped_order = event_to_sud.get_tag_and_order()
+                            type = 1   #### H1 type
+                        else:
+                            type = 2    #### H2 type                 
+            elif (len(buff_event) == nexternal): # is an S-event
+                    event_to_sud = buff_event
+                    n_part = nexternal 
+                    mapped_tag, mapped_order = event_to_sud.get_tag_and_order()
+                    type = 0   ### S type
             else:
-                nhel = 0
-            misc.sprint(nhel, Pdir, hel_dict)                        
-            raise Exception("Invalid matrix element for original computation (weight=0)")
+                logger.critical('ERROR: neither H nor S event!')
+                logger.critical(buff_event)
+                sys.exit(2)
+                
+            # Boost to partonic CM frame if not already in one for the momentum reshuffling 
+            E, px, py, pz = 0.,0.,0.,0.
```

Comment:
```suggestion
            p = lhe_parser.FourMomentum()
```
Note you can also at the top of the file doe 
```
from lhe_parser import FourMomentum
```
to simplify many of those line